### DISCUSSION TITLE
Array slices do not work with variables when their value is negative in C

### DISCUSSION THREAD

**Discussion by jalalium** [pyccel/pyccel#1313] [Q&A] [answered]
### Describe the bug
Array slices do not work as intended with variables that hold negative values in C.
`a[-1:]` does not give the same result as `a[c:]` when `c == -1`

### To Reproduce
Provide code to reproduce the behavior:
```python
import numpy as np

if __name__ == '__main__':
    a = [1,2,3]
    c = -1
    print(a[-c:])
```

### Error details
Provide the generated code, or the error message:
```c
#include "test.h"
#include <stdlib.h>
#include "ndarrays.h"
#include <stdint.h>
#include <string.h>
#include <stdio.h>
#include <inttypes.h>
int main()
{
    t_ndarray a = {.shape = NULL};
    int64_t c;
    int64_t i;
    a = array_create(1, (int64_t[]){INT64_C(3)}, nd_int64, false);
    int64_t array_dummy[] = {INT64_C(1), INT64_C(2), INT64_C(3)};
    memcpy(a.nd_int64, array_dummy, a.buffer_size);
    c = -INT64_C(1);
    printf("%s", "[");
    for (i = INT64_C(0); i < INT64_C(3) - c - INT64_C(1); i += INT64_C(1))
    {
        printf("%"PRId64" ", GET_ELEMENT(a, nd_int64, (int64_t)c + i));
    }
    printf("%"PRId64"]\n", GET_ELEMENT(a, nd_int64, (int64_t)c + (INT64_C(3) - c - INT64_C(1))));
    return 0;
}
```

### Expected behavior
The expected value is `[3]` but we get `[33 1 2 3]`

### Language
C

**Comment by yguclu** [comment, marked as answer]
This is to be expected, and it is explained in the [documentation](https://github.com/pyccel/pyccel/blob/261c1d9027efc3e7a2c33b6a1e7c081c48f0a602/tutorial/decorators.md#allow-negative-index).

We handle negative literal integers by default, but if the index is a variable (or an expression) we assume that it is always positive. Doing otherwise would require printing a lot of code to handle the two cases (positive vs. negative index), which is useless in 99% of the situations.

If you want to allow for negative indices through variables and expressions you have to pass the decorator `@allow_negative_index('a')`.
